[SimplifyCFG] Relax `cttz` cost check in `simplifySwitchOfPowersOfTwo`

antoniofrighetto opened the pull request:
We should be able to allow `simplifySwitchOfPowersOfTwo` transform to take place, as, on recent X86 targets, the weighted latency-size appears to be 2. This favours computing trailing zeroes and indexing into a smaller value table, over generating a jump table with an indirect branch (which overall should be more efficient).

antoniofrighetto commented:
Okay, I think we miss some non-zero power-of-two checks here...

nikic reviewed (commented):
I think relaxing this to 2 is fine. Can you add an x86 test?

antoniofrighetto commented:
Added, thanks, shall we take the opportunity extend this to cover reachable default cases too, or better this done separately?

nikic commented:
> Added, thanks, shall we take the opportunity extend this to cover reachable default cases too, or better this done separately?

Separately.

dtcxzyw reviewed (approved):
LG

nikic reviewed (approved):
LGTM

One thing this regressions is a power of two identity mapping (1 => 1, 2 => 2, 4 => 4, 8 => 8), which previously would just simplify to the input and now instead becomes cttz + table lookup. See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/2486/files/7b32e72e21d0f2d7df01d737f7a43222df4b1b60#diff-469c9a904297396b18fb7e4fd311b56e92874a0c79a13d5f960f01304ec305de.

It probably make sense to not perform this transform in the early SimplifyCFG runs.

RKSimon commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -7211,17 +7213,15 @@ static bool simplifySwitchOfPowersOfTwo(SwitchInst *SI, IRBuilder<> &Builder,
+  IntrinsicCostAttributes Attrs(Intrinsic::cttz, CondTy,
+                                {Condition, ConstantInt::getTrue(Context)});
+  if (TTI.getIntrinsicInstrCost(Attrs, TTI::TCK_SizeAndLatency) >
+      TTI::TCC_Basic * 2)
```
Is there not a cost for the alternative codegen to compare against? Using hardcoded TTI::TCC_* cost magic numbers tends to cause problems at some point.

antoniofrighetto committed:
[SimplifyCFG] Relax `cttz` cost check in `simplifySwitchOfPowersOfTwo`

We should be able to allow `simplifySwitchOfPowersOfTwo` transform
to take place, as, on recent X86 targets, the weighted latency-size
appears to be 2. This favours computing trailing zeroes and indexing
into a smaller value table, over generating a jump table with an
indirect branch, which overall should be more efficient.
